text/0069-schema-libraries.md

```diff
@@ -0,0 +1,169 @@
+is naturally declared as a Cedar entity type.
+With this RFC, anyone could provide a "library" representing Cedar definitions for
+OIDC types, and provide that library as a Cedar schema file at some URL; and then
+other Cedar users could use those definitions simply by importing the file from that URL.
+This would allow the Cedar community to gradually coalesce on the "best" way to
+represent an OIDC user in Cedar.
+
+### Motivations common to both scenarios
+
+In both of the above scenarios (within-organization and cross-organization),
+we obtain three key benefits:
+1. Saving each Cedar user the effort of writing common declarations themselves.
+  This facilitates code reuse in schemas, and makes it easier to get started
+  with Cedar.
+2. Providing a way to define common types and actions in a centralized way,
+  which ensures many schemas agree on the "correct" or "best practices"
+  definitions, and provides a single place to make updates if updates are
+  required.
+3. Facilitating code reuse for Cedar authorization calls, not just schemas.
+  When everyone shares a common definition of `OIDC::User`, the community could
+  conceivably converge on a reusable library function for, e.g., converting an
+  OIDC token into Cedar entity data.
+  This would further make it easier to get started with Cedar.
+  (Note that this RFC does not propose the Cedar team writing or maintaining
+  either library definitions for use in schemas or library functions for use in
+  Cedar authorization calls. It only points out that the community could
+  converge on these things.)
+
+## Detailed design
+
+Import statements are only allowed at the top level, outside of all namespace
+declarations.
+(In the future, another RFC could propose allowing imports in other positions.)
+
+The target of the import must be a raw file containing a valid Cedar schema.
+This schema may contain any definitions that are valid today in Cedar schemas,
+including namespaces, entity type definitions, common type definitions, and
+action declarations.
+The validator will essentially concatenate all of these definitions into the
+schema at the location of the `import` statement.
+
+Cedar will autodetect whether the imported schema is a human-format or
+JSON-format schema.
+(Today, there are no strings that are both valid human-format and valid
+JSON-format schemas; this RFC proposes encoding that as a design principle in
+perpetuity.)
+In particular, all valid JSON-format schemas must have `{` as their first
+non-whitespace character, and no valid human-format schemas have `{` as their
+first non-whitespace character.
+
+This RFC does not propose any mechanism for versioning libraries.
+Instead, it proposes that versioning would be done _above_ the Cedar layer,
+i.e., should be the responsibility of library authors.
+For instance, library authors could provide a different URL for different
+versions of their library, avoiding changing the contents of the URL for the
```

Comment:
We should provide something akin to [sub resource integrity](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity#subresource_integrity_with_the_script_element) so users can guard against a changing import.

Reply:
Bringing here the summary of an offline conversation -- it makes a lot of sense to have some feature like this, but since including hashes directly in the schema file is a little ugly, maybe we'd want a separate lockfile, and at that point we might want to consider some separate Cedar.toml file (analogous to Cargo.toml) where you could declare all your dependencies, and possibly other global configuration.  Does the long-term path see Cedar having something more and more equivalent to `cargo`, complete with dependency/lockfile manager, package repository, etc?